File: Sources/AgentTrainer/Storage/InputEventFile.swift
import CoreGraphics
import Foundation

final class InputEventWriter: @unchecked Sendable {
    private let handle: FileHandle
    private let lock = NSLock()
    private var buffer = Data()
    private(set) var count = 0
    private var closed = false
    private var failure: Error?

    init(url: URL) throws {
        FileManager.default.createFile(atPath: url.path, contents: nil)
        handle = try FileHandle(forWritingTo: url)
        try handle.truncate(atOffset: 0)
        try handle.write(contentsOf: Data("ATREVT01".utf8))
        var version = UInt32(1).littleEndian
        try withUnsafeBytes(of: &version) { try handle.write(contentsOf: Data($0)) }
        buffer.reserveCapacity(256 * 1024)
    }

    func append(_ event: InputSample) {
        lock.lock()
        defer { lock.unlock() }
        guard !closed, failure == nil else { return }
        buffer.appendInteger(event.timestampNanos)
        buffer.append(event.kind.rawValue)
        buffer.append(event.isDown ? 1 : 0)
        buffer.append(event.button)
        buffer.append(0)
        buffer.appendInteger(event.keyCode)
        buffer.appendInteger(UInt16(0))
        buffer.appendInteger(event.modifiers)
        buffer.appendDouble(event.x)
        buffer.appendDouble(event.y)
        buffer.appendDouble(event.deltaX)
        buffer.appendDouble(event.deltaY)
        buffer.appendDouble(event.scrollX)
        buffer.appendDouble(event.scrollY)
        count += 1
        if buffer.count >= 256 * 1024 { flushLocked() }
    }

    func finish() throws -> Int {
        lock.lock()
        defer { lock.unlock() }
        if closed {
            if let failure { throw failure }
            return count
        }
        flushLocked()
        if failure == nil {
            do { try handle.synchronize() }
            catch { failure = error }
        }
        do { try handle.close() }
        catch { if failure == nil { failure = error } }
        closed = true
        if let failure {
            throw AgentTrainerError.storage("The recorded input file could not be saved: \(failure.localizedDescription)")
        }
        return count
    }

    private func flushLocked() {
        guard !buffer.isEmpty, failure == nil else { return }
        do { try handle.write(contentsOf: buffer) }
        catch { failure = error }
        buffer.removeAll(keepingCapacity: true)
    }
}

enum InputEventReader {
    static let recordSize = 72

    /// Memory-mapped random-access view used by dataset construction. Large,
    /// mouse-heavy recordings no longer expand every fixed-width record into a
    /// heap-allocated Swift value before the first training sample is written.
    struct MappedEvents: RandomAccessCollection {
        typealias Index = Int
        typealias Element = InputSample

        fileprivate let data: Data
        let startIndex = 0
        let endIndex: Int

        fileprivate init(data: Data) throws {
            try InputEventReader.validate(data)
            self.data = data
            endIndex = (data.count - 12) / InputEventReader.recordSize
        }

        subscript(position: Int) -> InputSample {
            precondition(indices.contains(position))
            return InputEventReader.decodeRecord(in: data, index: position)
        }
    }

    struct MouseDiagnostics: Sendable, Equatable {
        var moveEventCount = 0
        var nonzeroDeltaCount = 0
        var absolutePositionChangeCount = 0
        var outOfCaptureBoundsCount = 0
        var accumulatedDeltaMagnitude = 0.0
        var maximumDeltaMagnitude = 0.0

        var nonzeroDeltaFraction: Double { Double(nonzeroDeltaCount) / Double(max(1, moveEventCount)) }
        var absolutePositionChangeFraction: Double { Double(absolutePositionChangeCount) / Double(max(1, moveEventCount - 1)) }
        var meanActiveDeltaMagnitude: Double { accumulatedDeltaMagnitude / Double(max(1, nonzeroDeltaCount)) }
        var isGameCamera: Bool {
            moveEventCount >= 20 && nonzeroDeltaCount > 0 && absolutePositionChangeFraction < 0.05
        }
        var positionsAreValid: Bool { outOfCaptureBoundsCount == 0 }
    }

    struct Summary: Sendable {
        var preview: [InputSample]
        var usedKeyCodes: Set<UInt16>
        var keyEventCount: Int
        var mouseEventCount: Int
        var mouse = MouseDiagnostics()
    }

    static func read(url: URL) throws -> [InputSample] {
        let mapped = try mapped(url: url)
        var result: [InputSample] = []
        result.reserveCapacity(mapped.count)
        result.append(contentsOf: mapped)
        return result
    }

    static func mapped(url: URL) throws -> MappedEvents {
        try MappedEvents(data: Data(contentsOf: url, options: .mappedIfSafe))
    }

    static func summarize(url: URL, previewLimit: Int = 80, globalRect: CGRect? = nil) throws -> Summary {
        let data = try Data(contentsOf: url, options: .mappedIfSafe)
        var summary = Summary(preview: [], usedKeyCodes: [], keyEventCount: 0, mouseEventCount: 0)
        var previousMousePosition: CGPoint?
        summary.preview.reserveCapacity(max(0, previewLimit))
        try forEach(in: data) { event in
            if summary.preview.count < previewLimit { summary.preview.append(event) }
            switch event.kind {
            case .key:
                summary.keyEventCount += 1
                summary.usedKeyCodes.insert(event.keyCode)
            case .flags:
                summary.usedKeyCodes.formUnion(modifierKeyCodes(in: event.modifiers))
            case .mouseMove:
                summary.mouseEventCount += 1
                summary.mouse.moveEventCount += 1
                let magnitude = abs(event.deltaX) + abs(event.deltaY)
                if magnitude > 0 {
                    summary.mouse.nonzeroDeltaCount += 1
                    summary.mouse.accumulatedDeltaMagnitude += magnitude
                    summary.mouse.maximumDeltaMagnitude = max(summary.mouse.maximumDeltaMagnitude, magnitude)
                }
                let position = CGPoint(x: event.x, y: event.y)
                if let previousMousePosition,
                   abs(position.x - previousMousePosition.x) > 0.01 || abs(position.y - previousMousePosition.y) > 0.01 {
                    summary.mouse.absolutePositionChangeCount += 1
                }
                previousMousePosition = position
                if let globalRect, !globalRect.insetBy(dx: -0.5, dy: -0.5).contains(position) {
                    summary.mouse.outOfCaptureBoundsCount += 1
                }
            case .mouseButton, .scroll: summary.mouseEventCount += 1
            }
        }
        return summary
    }

    static func demonstratedKeyCodes(url: URL) throws -> Set<UInt16> {
        let data = try Data(contentsOf: url, options: .mappedIfSafe)
        var result: Set<UInt16> = []
        try forEach(in: data) { event in
            if event.kind == .key { result.insert(event.keyCode) }
            result.formUnion(modifierKeyCodes(in: event.modifiers))
        }
        return result
    }

    static func mouseDiagnostics(events: [InputSample], globalRect: CGRect? = nil) -> MouseDiagnostics {
        var diagnostics = MouseDiagnostics()
        var previousPosition: CGPoint?
        for event in events where event.kind == .mouseMove {
            diagnostics.moveEventCount += 1
            let magnitude = abs(event.deltaX) + abs(event.deltaY)
            if magnitude > 0 {
                diagnostics.nonzeroDeltaCount += 1
                diagnostics.accumulatedDeltaMagnitude += magnitude
                diagnostics.maximumDeltaMagnitude = max(diagnostics.maximumDeltaMagnitude, magnitude)
            }
            let position = CGPoint(x: event.x, y: event.y)
            if let previousPosition,
               abs(position.x - previousPosition.x) > 0.01 || abs(position.y - previousPosition.y) > 0.01 {
                diagnostics.absolutePositionChangeCount += 1
            }
            previousPosition = position
            if let globalRect, !globalRect.insetBy(dx: -0.5, dy: -0.5).contains(position) {
                diagnostics.outOfCaptureBoundsCount += 1
            }
        }
        return diagnostics
    }

    private static func modifierKeyCodes(in flags: UInt64) -> Set<UInt16> {
        let mappings: [(CGEventFlags, UInt16)] = [(.maskShift, 56), (.maskControl, 59), (.maskAlternate, 58), (.maskCommand, 55)]
        return Set(mappings.compactMap { flags & $0.0.rawValue != 0 ? $0.1 : nil })
    }

    private static func validate(_ data: Data) throws {
        guard data.count >= 12, String(data: data.prefix(8), encoding: .utf8) == "ATREVT01" else { throw AgentTrainerError.storage("Invalid AgentTrainer input event file.") }
        var headerCursor = 8
        let version: UInt32 = data.readInteger(at: &headerCursor)
        guard version == 1, (data.count - 12).isMultiple(of: recordSize) else {
            throw AgentTrainerError.storage("This AgentTrainer input event file is unsupported or incomplete.")
        }
        let count = (data.count - 12) / recordSize
        var previousTimestamp: UInt64?
        for index in 0..<count {
            let recordStart = 12 + index * recordSize
            let rawKind = data[recordStart + 8]
            guard InputEventKind(rawValue: rawKind) != nil else {
                throw AgentTrainerError.storage("This AgentTrainer input event file contains an unknown event kind.")
            }
            var cursor = recordStart
            let timestamp: UInt64 = data.readInteger(at: &cursor)
            if let previousTimestamp, timestamp < previousTimestamp {
                throw AgentTrainerError.storage("This AgentTrainer input event file is not ordered by capture time.")
            }
            previousTimestamp = timestamp
            let event = decodeRecord(in: data, index: index)
            guard [event.x, event.y, event.deltaX, event.deltaY, event.scrollX, event.scrollY].allSatisfy(\.isFinite) else {
                throw AgentTrainerError.storage("This AgentTrainer input event file contains a non-finite control value.")
            }
        }
    }

    private static func decodeRecord(in data: Data, index: Int) -> InputSample {
        var cursor = 12 + index * recordSize
        let timestamp: UInt64 = data.readInteger(at: &cursor)
        let kindRaw = data[cursor]; cursor += 1
        let isDown = data[cursor] != 0; cursor += 1
        let button = data[cursor]; cursor += 2
        let keyCode: UInt16 = data.readInteger(at: &cursor); cursor += 2
        let modifiers: UInt64 = data.readInteger(at: &cursor)
        let x = data.readDouble(at: &cursor), y = data.readDouble(at: &cursor)
        let dx = data.readDouble(at: &cursor), dy = data.readDouble(at: &cursor)
        let sx = data.readDouble(at: &cursor), sy = data.readDouble(at: &cursor)
        // Validation above guarantees the closed version-1 event kind set.
        let kind = InputEventKind(rawValue: kindRaw)!
        return InputSample(timestampNanos: timestamp, kind: kind, x: x, y: y, deltaX: dx, deltaY: dy, button: button, scrollX: sx, scrollY: sy, keyCode: keyCode, modifiers: modifiers, isDown: isDown)
    }

    private static func forEach(in data: Data, _ body: (InputSample) -> Void) throws {
        try validate(data)
        let count = (data.count - 12) / recordSize
        for index in 0..<count {
            body(decodeRecord(in: data, index: index))
        }
    }
}

private extension Data {
    mutating func appendInteger<T: FixedWidthInteger>(_ value: T) {
        var little = value.littleEndian
        Swift.withUnsafeBytes(of: &little) { append(contentsOf: $0) }
    }

    mutating func appendDouble(_ value: Double) {
        appendInteger(value.bitPattern)
    }

    func readInteger<T: FixedWidthInteger>(at cursor: inout Int) -> T {
        let size = MemoryLayout<T>.size
        let value = self[cursor..<(cursor + size)].withUnsafeBytes { $0.loadUnaligned(as: T.self) }
        cursor += size
        return T(littleEndian: value)
    }

    func readDouble(at cursor: inout Int) -> Double {
        Double(bitPattern: readInteger(at: &cursor) as UInt64)
    }
}

File: WindowsRecorder/src/AgentTrainer.Recorder.Core/InputEventFile.cs
using System.Buffers;
using System.Buffers.Binary;
using System.Text;

namespace AgentTrainer.Recorder.Core;

public sealed class InputEventWriter : IDisposable
{
    public static ReadOnlySpan<byte> Magic => "ATREVT01"u8;
    public const uint FormatVersion = 1;
    public const int HeaderSize = 12;
    public const int RecordSize = 72;
    private const int BufferCapacity = 256 * 1024;

    private readonly FileStream _stream;
    private readonly byte[] _buffer;
    private readonly object _gate = new();
    private int _bufferCount;
    private bool _closed;
    private ulong? _lastTimestamp;

    public InputEventWriter(string path)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(path))!);
        _stream = new FileStream(path, FileMode.CreateNew, FileAccess.Write, FileShare.Read, 4096,
            FileOptions.SequentialScan | FileOptions.WriteThrough);
        _buffer = ArrayPool<byte>.Shared.Rent(BufferCapacity + RecordSize);
        _stream.Write(Magic);
        Span<byte> version = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(version, FormatVersion);
        _stream.Write(version);
    }

    public int Count { get; private set; }

    public void Append(InputSample sample)
    {
        lock (_gate)
        {
            ObjectDisposedException.ThrowIf(_closed, this);
            if (_lastTimestamp is { } previous && sample.TimestampNanos < previous)
                throw new InvalidDataException("Input timestamps must be nondecreasing.");
            if (!Enum.IsDefined(sample.Kind) || !AllFinite(sample))
                throw new InvalidDataException("Input sample contains an unsupported kind or non-finite value.");

            Span<byte> record = stackalloc byte[RecordSize];
            record.Clear();
            BinaryPrimitives.WriteUInt64LittleEndian(record, sample.TimestampNanos);
            record[8] = (byte)sample.Kind;
            record[9] = sample.IsDown ? (byte)1 : (byte)0;
            record[10] = sample.Button;
            BinaryPrimitives.WriteUInt16LittleEndian(record[12..], sample.KeyCode);
            BinaryPrimitives.WriteUInt64LittleEndian(record[16..], sample.Modifiers);
            WriteDouble(record[24..], sample.X);
            WriteDouble(record[32..], sample.Y);
            WriteDouble(record[40..], sample.DeltaX);
            WriteDouble(record[48..], sample.DeltaY);
            WriteDouble(record[56..], sample.ScrollX);
            WriteDouble(record[64..], sample.ScrollY);
            record.CopyTo(_buffer.AsSpan(_bufferCount));
            _bufferCount += RecordSize;
            Count++;
            _lastTimestamp = sample.TimestampNanos;
            if (_bufferCount >= BufferCapacity) FlushBuffer();
        }
    }

    public int Finish()
    {
        lock (_gate)
        {
            if (_closed) return Count;
            FlushBuffer();
            _stream.Flush(flushToDisk: true);
            _stream.Dispose();
            _closed = true;
            ArrayPool<byte>.Shared.Return(_buffer);
            return Count;
        }
    }

    public void Dispose() => Finish();

    private void FlushBuffer()
    {
        if (_bufferCount == 0) return;
        _stream.Write(_buffer, 0, _bufferCount);
        _bufferCount = 0;
    }

    private static bool AllFinite(InputSample value) =>
        double.IsFinite(value.X) && double.IsFinite(value.Y)
        && double.IsFinite(value.DeltaX) && double.IsFinite(value.DeltaY)
        && double.IsFinite(value.ScrollX) && double.IsFinite(value.ScrollY);

    private static void WriteDouble(Span<byte> destination, double value) =>
        BinaryPrimitives.WriteInt64LittleEndian(destination, BitConverter.DoubleToInt64Bits(value));
}

public sealed record InputEventSummary(
    int Count,
    int KeyEventCount,
    int MouseEventCount,
    SortedSet<ushort> UsedKeyCodes,
    IReadOnlyList<InputSample> Preview,
    MouseDiagnostics Mouse,
    InputSample? First,
    InputSample? Last);

public sealed record MouseDiagnostics
{
    public int MoveEventCount { get; init; }
    public int NonzeroDeltaCount { get; init; }
    public int AbsolutePositionChangeCount { get; init; }
    public int OutOfCaptureBoundsCount { get; init; }
    public double AccumulatedDeltaMagnitude { get; init; }
    public double MaximumDeltaMagnitude { get; init; }
    public double NonzeroDeltaFraction => (double)NonzeroDeltaCount / Math.Max(1, MoveEventCount);
    public double AbsolutePositionChangeFraction => (double)AbsolutePositionChangeCount / Math.Max(1, MoveEventCount - 1);
    public double MeanActiveDeltaMagnitude => AccumulatedDeltaMagnitude / Math.Max(1, NonzeroDeltaCount);
    public bool IsGameCamera => MoveEventCount >= 20 && NonzeroDeltaCount > 0 && AbsolutePositionChangeFraction < 0.05;
    public bool PositionsAreValid => OutOfCaptureBoundsCount == 0;
}

public static class InputEventReader
{
    public static InputEventSummary Summarize(string path, int previewLimit = 80, CodableRect? globalRect = null)
    {
        using var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, 128 * 1024, FileOptions.SequentialScan);
        ValidateHeaderAndLength(stream);
        var count = checked((int)((stream.Length - InputEventWriter.HeaderSize) / InputEventWriter.RecordSize));
        var preview = new List<InputSample>(Math.Min(Math.Max(0, previewLimit), count));
        var keys = new SortedSet<ushort>();
        var keyEvents = 0;
        var mouseEvents = 0;
        var moveCount = 0;
        var nonzeroDelta = 0;
        var positionChanges = 0;
        var outside = 0;
        var accumulated = 0.0;
        var maximum = 0.0;
        (double X, double Y)? previousPosition = null;
        InputSample? first = null;
        InputSample? last = null;
        ulong? priorTimestamp = null;
        var recordBytes = new byte[InputEventWriter.RecordSize];

        for (var index = 0; index < count; index++)
        {
            stream.ReadExactly(recordBytes);
            var sample = Decode(recordBytes);
            if (priorTimestamp is { } prior && sample.TimestampNanos < prior)
                throw new InvalidDataException("AgentTrainer input events are not ordered by capture time.");
            priorTimestamp = sample.TimestampNanos;
            first ??= sample;
            last = sample;
            if (preview.Count < previewLimit) preview.Add(sample);
            switch (sample.Kind)
            {
                case InputEventKind.Key:
                    keyEvents++;
                    keys.Add(sample.KeyCode);
                    break;
                case InputEventKind.Flags:
                    AddModifierKeys(keys, sample.Modifiers);
                    break;
                case InputEventKind.MouseMove:
                    mouseEvents++;
                    moveCount++;
                    var magnitude = Math.Abs(sample.DeltaX) + Math.Abs(sample.DeltaY);
                    if (magnitude > 0)
                    {
                        nonzeroDelta++;
                        accumulated += magnitude;
                        maximum = Math.Max(maximum, magnitude);
                    }
                    if (previousPosition is { } point &&
                        (Math.Abs(sample.X - point.X) > 0.01 || Math.Abs(sample.Y - point.Y) > 0.01)) positionChanges++;
                    previousPosition = (sample.X, sample.Y);
                    if (globalRect is { } rect && !ContainsWithTolerance(rect, sample.X, sample.Y)) outside++;
                    break;
                case InputEventKind.MouseButton:
                case InputEventKind.Scroll:
                    mouseEvents++;
                    break;
            }
        }

        return new InputEventSummary(count, keyEvents, mouseEvents, keys, preview,
            new MouseDiagnostics
            {
                MoveEventCount = moveCount,
                NonzeroDeltaCount = nonzeroDelta,
                AbsolutePositionChangeCount = positionChanges,
                OutOfCaptureBoundsCount = outside,
                AccumulatedDeltaMagnitude = accumulated,
                MaximumDeltaMagnitude = maximum
            }, first, last);
    }

    public static IReadOnlyList<InputSample> ReadAll(string path) => Summarize(path, int.MaxValue).Preview;

    private static void ValidateHeaderAndLength(FileStream stream)
    {
        if (stream.Length < InputEventWriter.HeaderSize ||
            (stream.Length - InputEventWriter.HeaderSize) % InputEventWriter.RecordSize != 0)
            throw new InvalidDataException("AgentTrainer input event file is incomplete.");
        Span<byte> header = stackalloc byte[InputEventWriter.HeaderSize];
        stream.ReadExactly(header);
        if (!header[..8].SequenceEqual(InputEventWriter.Magic) ||
            BinaryPrimitives.ReadUInt32LittleEndian(header[8..]) != InputEventWriter.FormatVersion)
            throw new InvalidDataException("AgentTrainer input event header is unsupported.");
    }

    private static InputSample Decode(ReadOnlySpan<byte> record)
    {
        var rawKind = record[8];
        if (!Enum.IsDefined(typeof(InputEventKind), rawKind))
            throw new InvalidDataException("AgentTrainer input event file contains an unknown event kind.");
        var sample = new InputSample(
            BinaryPrimitives.ReadUInt64LittleEndian(record),
            (InputEventKind)rawKind,
            ReadDouble(record[24..]), ReadDouble(record[32..]),
            ReadDouble(record[40..]), ReadDouble(record[48..]),
            record[10], ReadDouble(record[56..]), ReadDouble(record[64..]),
            BinaryPrimitives.ReadUInt16LittleEndian(record[12..]),
            BinaryPrimitives.ReadUInt64LittleEndian(record[16..]),
            record[9] != 0);
        if (!new[] { sample.X, sample.Y, sample.DeltaX, sample.DeltaY, sample.ScrollX, sample.ScrollY }.All(double.IsFinite))
            throw new InvalidDataException("AgentTrainer input event file contains a non-finite control value.");
        return sample;
    }

    private static double ReadDouble(ReadOnlySpan<byte> value) =>
        BitConverter.Int64BitsToDouble(BinaryPrimitives.ReadInt64LittleEndian(value));

    private static void AddModifierKeys(SortedSet<ushort> result, ulong flags)
    {
        if ((flags & QuartzModifierFlags.Shift) != 0) result.Add(56);
        if ((flags & QuartzModifierFlags.Control) != 0) result.Add(59);
        if ((flags & QuartzModifierFlags.Option) != 0) result.Add(58);
        if ((flags & QuartzModifierFlags.Command) != 0) result.Add(55);
    }

    private static bool ContainsWithTolerance(CodableRect rect, double x, double y) =>
        x >= rect.X - 0.5 && y >= rect.Y - 0.5
        && x <= rect.X + rect.Width + 0.5 && y <= rect.Y + rect.Height + 0.5;
}
